classes/Profile.php

```diff
@@ -191,44 +189,50 @@ public static function resetCacheAccesses()
 
     /**
      * @param int $idProfile Profile ID
-     * @param string $type Type
-     * @param array $cacheData Cached data
+     * @param array $defaultData Cached data
+     * @param array $accesses Data loaded from the database
      */
-    private static function fillCacheAccesses($idProfile, $type, $cacheData = [])
+    private static function fillCacheAccesses($idProfile, $defaultData = [], $accesses = [])
     {
         foreach (Tab::getTabs(Context::getContext()->language->id) as $tab) {
-            self::$_cache_accesses[$idProfile][$type][$tab[$type]] = array_merge(
-                array(
-                    'id_tab' => $tab['id_tab'],
-                    'class_name' => $tab['class_name'],
-                ),
-                $cacheData
-            );
+            $accessData = [];
+            if (isset($accesses[strtoupper($tab['class_name'])])) {
+                $accessData = $accesses[strtoupper($tab['class_name'])];
+            }
+
+            foreach (self::ALLOWED_PROFILE_TYPE_CHECK as $type) {
+                self::$_cache_accesses[$idProfile][$type][$tab[$type]] = array_merge(
+                    array(
+                        'id_tab' => $tab['id_tab'],
+                        'class_name' => $tab['class_name'],
+                    ),
+                    $defaultData,
+                    $accessData
+                );
+            }
         }
     }
 
     /**
-     * Find tab type information by authorization slug.
+     * Creates the array of accesses [role => add / view / edit / delete] from a given list of roles
      *
-     * @param string $type
-     * @param string $authSlug
+     * @param array $rolesGiven
      *
-     * @return int
+     * @return array
      */
-    private static function findTabTypeInformationByAuthSlug($type, $authSlug)
+    private static function generateAccessesArrayFromPermissions($rolesGiven)
     {
-        preg_match(
-            '/ROLE_MOD_[A-Z]+_(?P<classname>[A-Z][A-Z0-9]*)_(?P<auth>[A-Z]+)/',
-            $authSlug,
-            $matches
-        );
-
-        $result = Db::getInstance()->getRow('
-            SELECT `' . $type . '`
-            FROM `' . _DB_PREFIX_ . 'tab` t
-            WHERE UCASE(`class_name`) = "' . $matches['classname'] . '"
-        ');
+        // Modify array to merge the class names together.
+        $accessPerTab = [];
+        foreach ($rolesGiven as $role) {
+            preg_match(
+                '/ROLE_MOD_[A-Z]+_(?P<classname>[A-Z][A-Z0-9]*)_[A-Z]+/',
+                $role['slug'],
+                $matches
+            );
+            $accessPerTab[$matches['classname']][array_search('1', $role)] = '1';
```

Comment:
no checking of the `preg_match` `$matches` result?

Reply:
This value came from database, and was the old check.

Reply:
There was no such check on the previous commit.

What do you suggest?

```php
if (empty($matches['classname'])) {
    continue;
}
```
?

Reply:
out of safety yes

Reply:
Done, you may review again. :)

Reply:
@Quetzacoalt91 it's been approved!